scMerlin 2.5.42 Release (Dual-Wan Support)

ExtremeFiretop opened the pull request:
scMerlin 2.5.42 Release (Dual-Wan Support)

ExtremeFiretop commented:
> > @Martinski4GitHub
> > Can you give me a thumbs up or thumbs down on merging when possible.
> > Thank you! :)
> 
> I reviewed the code and made some fixes and improvements in some old code, and also in the latest code, just making things more robust.

Thank you Martinski! I'm still seeing some reports in the forums this morning so I'll have them re-test!

Martinski4GitHub commented on scmerlin.sh:
```diff
@@ -308,8 +308,8 @@ GetTemperatureString()
-    then printf "${REDct}%s${CLRct}" "$theTemptrVal"
-    else printf "${GRNct}%s°C${CLRct}" "$theTemptrVal"
+    then printf "${REDct}%s${CLRct}" "$1"
+    else printf "${GRNct}%s°C${CLRct}" "$1"
```
This was actually correct and as intended; the variable `theTemptrVal` has already been set in a previous function call and passed as the argument just to check if we have a temperature value or some error message. But,  I can see how it may appear wrong and confusing unless you follow the execution flow.

Martinski4GitHub commented on scmerlin.sh:
```diff
@@ -143,26 +143,26 @@ _GetWiFiBandsSupported_()
+        "RT-BE96U" | "GT-AXE11000")
+            Band_6G_1_Support=true
+            ;;
+    esac
```
>
> 1. case for _GetWiFiBandsSupported_() matches GT-BE98 in an earlier branch, so later 6 GHz flags never get set.
>
I don't understand what you're saying. Can you elaborate?
Each of the `IF` statements was independent of each other - there was no `ELSE` - so the logic flows through each `IF` statement. IOW, there was no "early exit" for each case.

ExtremeFiretop replied:
In a single case block I thought only the first matching arm runs, and the whole case ends at the ;; I thought right? 

So the script would scan arms top-to-bottom, executes the first pattern that matches, then jumps past the entire case after the arm’s.

Since "GT-BE98" appears in two arms, a GT-BE98 router will hit the first arm (sets Band_5G_2_Support=true) and won’t execute the second arm (so Band_6G_1_Support never gets set there). That’s not like “independent ifs”? Sorry if I'm confusing that point.

Martinski4GitHub replied:
> In a single case block I thought only the first matching arm runs, and the whole case ends at the ;; I thought right?
> 
You're looking at OLD code!!!!!  Which has already been fixed!!
How and why exactly are you reviewing OLD code?? 
Make sure 100% that you've got the very latest "develop' branch changes.

![ScMerlin_GitHub](https://github.com/user-attachments/assets/cbcf24a2-3f7c-4c04-aabc-ba7121cd95eb)

ExtremeFiretop replied:
> > In a single case block I thought only the first matching arm runs, and the whole case ends at the ;; I thought right?
> 
> You're looking at OLD code!!!!! Which has already been fixed!! How and why exactly are you reviewing OLD code?? 
> 

Ummm Great question?! LOL. All I'm doing is clicking on:

<img width="963" height="368" alt="image" src="https://github.com/user-attachments/assets/40397db5-6fe9-4864-a86f-f22f11c39ec6" />

And that's what I see? Did something go haywire? How is that not what you see? LOL

ExtremeFiretop replied:
According to the desktop app I synced as recently as 31 minutes ago with the dev branch without issue:

<img width="922" height="358" alt="image" src="https://github.com/user-attachments/assets/f4d2459b-3277-4806-a7b0-7558e5ce1167" />

I got all your changes included but this one somehow.

Martinski4GitHub replied:
> Ah I see if I click on your commit specifically I can see your fix:
> 
> #31
> 
> But I guess it didn't sync to my side somehow. So I re-fixed your fix LOL. For lack of better words. But how did all the other stuff sync down?
>
>
I don't know how you prepare to review the latest code, but I **always** refresh my own development branch **FIRST**. And then I review each commit, especially if there are many individual commits. Never had any problems of getting OLD code.

ExtremeFiretop replied:
I think what happened is a merge conflict happened because I had started fixing that area when it synced the changes you merged in (including that area already fixed). But on my side it kept my changes in that part of the code, but synced everything else that it could.

Martinski4GitHub replied:
> > > Maybe this helps show the 2 points:
> > > <img alt="image" width="2000" height="1052" src="https://private-user-images.githubusercontent.com/1971404/486485774-d7a8ce4a-9a70-464a-95c1-94c892708969.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcyMjQ5ODYsIm5iZiI6MTc1NzIyNDY4NiwicGF0aCI6Ii8xOTcxNDA0LzQ4NjQ4NTc3NC1kN2E4Y2U0YS05YTcwLTQ2NGEtOTVjMS05NGM4OTI3MDg5NjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDdUMDU1ODA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzJjYTM5ODVhOTkzY2MzZmFjODdjNGVlOGViNmM4NmQ1MjZhNzg3ZDdmYzYxMzYyM2Y5ZDdjNmQ5Njk0YmYwMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hrqdYJmmrVrxQ4ynJa_z1LbzKIg1N6hrf0agJVE84OQ">
> > 
> > 
> > Nope, somehow you're still looking at OLD code!! Make sure to refresh and get the **LATEST** development branch **BEFORE** reviews!!!!
> 
> Yeah I'm convinced something went haywire, I got all your fixes synced by the Github desktop app, everything synced but this for some reason. Because I can guarantee I did not manually copy and paste all these fixes over.
>
>
Yeah, I don't know what happened on your end, so I cannot really advise on how to avoid it in the future. But it looks like you have a flaw in your own process to get the very latest code for review.

ExtremeFiretop replied:
Usually a merge conflict should ask for an action/resolution before merging, but I don't remember seeing any such screen to resolve any conflicts on my side.

ExtremeFiretop replied:
> > If you look at my commit you can clearly see I'm not undoing all your changes/fixes from your commit: [4dd022a](https://github.com/AMTM-OSR/scMerlin/commit/4dd022a22132acfbe79b31e94b927ca8fc33ad67)
> > So if I'm not undoing them, it's because I was in-sync, at least partially in sync. How that happened not sure.
> 
> Actually, that link shows that you **were** undoing my previous fixes. Here's a partial screenshot from that commit link:
> 
> ![ScMerlin_GitHub_02](https://private-user-images.githubusercontent.com/119833648/486487829-ca7a583d-6d94-4137-9120-bca01071905f.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcyMjcyNjQsIm5iZiI6MTc1NzIyNjk2NCwicGF0aCI6Ii8xMTk4MzM2NDgvNDg2NDg3ODI5LWNhN2E1ODNkLTZkOTQtNDEzNy05MTIwLWJjYTAxMDcxOTA1Zi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwN1QwNjM2MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MjQ4ZThkMTA0N2RiNzcyMDMyOTFjNDFhZWU0OWM1ZmFkY2I2YzQzNjYxZmE5ODhjYzA4OGE0YWEwYWU0NzBhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.AYi8ea-ep7ZvSd7t3WIY3NFZeKijnPkVbMpG20_00DM)

That's tunnel vision; seeing what you want to see. That's the only thing that was undone.
What about EVERYTHING else:

<img width="977" height="1153" alt="image" src="https://github.com/user-attachments/assets/4d3f353a-150f-4b3f-b1ff-c5c028f8c9fc" />

Means I was partially in sync. Anyways it was undone.

ExtremeFiretop replied:
> Actually, that link shows that you **were** undoing my previous fixes. Here's a partial screenshot from that commit link:

fix* not fixes

ExtremeFiretop replied:
This proves I was partially in sync. And something went haywire. Likely a bad merge conflict of some kind.

Martinski4GitHub replied:
> > No need to get defensive and split hairs about "fix" or "fixes" since we both know what specific changes we're talking about.
> 
> I’m not trying to split hairs. My intent is clarity and accuracy. That’s something we’ve both valued, and I’m aiming to keep our communication precise about what changed and why.
> 
> > The rest has already been clarified.
> 
> On the “has been clarified” point: I didn’t initially see confirmation that the clarification landed. The first sign that we were converging was when you said:
> 
> > Yeah, I don't know what happened, but something went very wrong. Maybe you found a bug in GitHub, LOL!!!
> 
> Before that, a few replies suggested I was not grabbing the latest code and undoing fixes, e.g.:
> 
> > Nope, somehow you're still looking at OLD code!!
> 
> And:
> 
> > Nope. Keep in mind that I actually look at the source code **AFTER** refreshing my own **develop** branch.
> 
> Literally reading back the messages, the only message that showed any signs of clarification being received was second to last message indicating you might understand what happened, up until that point you even doubled down by saying:
> 
> > Actually, that link shows that you **were** undoing my previous fixes.
> 
> When the point was to show that I only undid one thing (either by mistake/without realizing) and my issue was not "grabbing the latest code" as you like to say.
> 
> > Now, you're getting defensive??? I'm not in the mood to argue about who's getting tunnel vision.
> 
> My goal in sharing the commit link was to show that only one chunk had been reverted (unintentionally), not that I wasn’t pulling latest.
> 
> For context, my workflow auto-syncs the repo roughly hourly and won’t let me submit unless I’m in sync. If I ever don’t have the latest, I’ll be the one to call it out explicitly. But let's not assume that to be the case as that will be wrong 95% of the time and is getting tiring to hear and defend as the default cause.
> 
> ... Likely root cause (what I found today)
> 
> While picking this back up, I realized GitHub Desktop lets you select/deselect specific hunks during merges/conflict resolution. It looks like I probably deselected that particular part by mistake:
> 
> Selected: <img alt="image" width="2000" height="855" src="https://private-user-images.githubusercontent.com/1971404/486906642-f070d6fd-08d4-49e7-b4c7-1a53c51daa3c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc0MDk3NzMsIm5iZiI6MTc1NzQwOTQ3MywicGF0aCI6Ii8xOTcxNDA0LzQ4NjkwNjY0Mi1mMDcwZDZmZC0wOGQ0LTQ5ZTctYjRjNy0xYTUzYzUxZGFhM2MucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDlUMDkxNzUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Zjg4OGU2YzliMDZlZjFlMmE0MGNhNzljMjMyNGIwODJkNWZiNjBhZDBhZTAxN2Y4MTMzNjdjYTFhNTNjOTZhYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.5XVPNqHwxv2rcrzhcMvC3en_utJ4ui2ZqUCxb1QGBXU">
> 
> Deselected: <img alt="image" width="2000" height="881" src="https://private-user-images.githubusercontent.com/1971404/486906688-f86494ac-523c-429e-bbdc-62b3ca86511f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc0MDk3NzMsIm5iZiI6MTc1NzQwOTQ3MywicGF0aCI6Ii8xOTcxNDA0LzQ4NjkwNjY4OC1mODY0OTRhYy01MjNjLTQyOWUtYmJkYy02MmIzY2E4NjUxMWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDlUMDkxNzUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTcxNDZhNTA4NjY2MWU2N2I1YWE3ODY4NmQ1OWY4NjVmN2YzZjYyZWQyNjhmM2EzY2ZlMzZjYmM2OGU1ZTkzZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.d9J_vYcaORGev9VR2RBlddSL3cX8JaEApojXHxIhenc">
> 
> Given this, it was likely a mis-click rather than a GitHub bug.
> 
> I didn't know this feature existed until 30 minutes ago, this is likely what happened, else I'll blame my cat for de-selecting it while walking across my keyboard. Either way, it could of been a laughing matter but it felt as you were clearly set on telling me I was in the wrong or did something wrong.
> 
> Going forward
> 
>     * Let’s assume I’m on latest **unless I say otherwise**.
> 
>     * If something looks off, I’ll double-check and report back with the exact commit/hunk so we can converge quickly.
> 
>     * I’ll be extra careful with selective staging in Desktop to avoid this particular footgun.
> 
> 
> Thanks for sticking with it. appreciate the collaboration. Let’s move on now that the cause is clear.
>
>
>
It's very clear that we were **not** talking about exactly the same things in our messages.

In my view, there were ***two*** problems with the OLD code using the CASE statements:
1) Not setting the var for one 6GHz band support (e.g. GT-BE98 router)
2) Not setting the var for 2nd 6GHz band support (e.g. GT-BE98_PRO router)

So technically and accurately, these were ***2*** bugs; therefore, there were ***2*** fixes (plural ***not*** singular) with my IF statement changes. That's why I wrote in my message:
>
> Actually, that link shows that you were **undoing my previous fixes.**
>
Note that I did ***NOT*** say "... you were undoing ***ALL*** my previous fixes" because I was referring only to the IF statement fixes - nothing else. And this point was further emphasized with my screenshot showing a statement in a yellow box with the red arrow pointing at the block of IF statement fixes:
>
> "**This is the LATESt code being undone with your commit**."
>
There was no indication at all of any other code being undone, so IMO there was no other code to discuss or talk about. So, the only remaining issue (at least for me) was how you ended up editing and modifying a shell script file that still had some of the OLD code, and then the commit link was clearly showing that you were actually overwriting the NEW code (the IF statement fixes) with your latest changes.

In your messages, however, you think of the IF statement changes as one single fix, which again, is not technically correct. But that was not a real issue to argue about until you brought it up.

Note that I had already acknowledged your fixes for "`Band_5G_2_Support`" variable (lower versus upper case '**S**'), and we had already talked about and settled the changes made in the '`GetTemperatureString()`' function. Therefore, in my view, those were no longer subject to further clarification or discussion, and that's why I said:
>
>
> ... The rest has already been clarified.
>
>
Because there were no other fixes to talk about since all other changes had already been reviewed, discussed, and clarified after we both had agreed on their resolution at that point.

Martinski4GitHub commented on scmerlin.sh:
```diff
@@ -113,7 +113,7 @@ readonly SUPPORTstr="$(nvram get rc_support)"
 Band_5G_1_Support=false
-Band_5G_2_support=false
+Band_5G_2_Support=false
 Band_6G_1_Support=false
```
> @Martinski4GitHub
> 
> I'm not actively looking for mistakes, I read it like a book, as long as I understand its a pass from me most of the time. But sometimes I run across them. My eye will never be as good as yours.
>
>
LOL!!! Yeah, that's why we need a 2nd pair of eyes, especially after my own are very tired after midnight :>).

ExtremeFiretop commented:
@Martinski4GitHub 

After some deeper investigation with [Stephen](https://www.snbforums.com/members/stephen-harrington.53476/) from the forums. It seems his reported issue actually has nothing to do with our code directly.

I tested by trying to setup one of my second routers in Dual-WAN with fail over and was unable to replicate his reports. Everything worked perfectly for me.

So I reached out to him via PM and made him provide me some nvram values, some system logs and run some functions, and it's clear to me that the issue is he is using another script for the Dual-WAN fail-over compared to the stock built-in fail-over function I was using.

In my tests everything was rock solid but I was using the built-in fail over function with the firmware, he had that disabled and was using this script: dual-wan-failover found here: https://www.snbforums.com/threads/dual-wan-failover-v2-release.83674/

In my tests I get a clear "connected" status in the system logs that triggers for both WAN interfaces as found below:

FAIL-OVER to WAN1
```
Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 init)
Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 connecting)
Sep  8 16:40:37 custom_script: Running /jffs/scripts/service-event-end (args: restart wan_if)
Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 disconnected)
Sep  8 16:40:38 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
Sep  8 16:40:39 custom_script: Running /jffs/scripts/service-event (args: restart wan_line)
Sep  8 16:40:39 custom_script: Running /jffs/scripts/wan-event (args: 1 connected)
Sep  8 16:40:39 custom_script: Running /jffs/scripts/firewall-start (args: eth4)
```

FALL-BACK to WAN0

```
Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 init)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 connecting)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/service-event-end (args: restart wan_if)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 disconnected)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 connected)
Sep  8 18:49:29 custom_script: Running /jffs/scripts/firewall-start (args: eth0)
```

However Stephen with the other script is only getting one "connected" state for WAN0 and nothing for WAN1.
He also confirmed the issue existed with the current production version now, this goes against his original report found in the forums here to you confirming it worked correctly: https://www.snbforums.com/threads/scmerlin-2-5-41-service-and-script-control-menu-for-asuswrt-merlin.89224/post-963668

So seems like this issue exists in production and was not introduced in the dev branch, I can confirm that all works correctly with the stock firmware fail-over and so can stephen.

So the question now is, is it worth troubleshooting further or call it a known issue/compatible conflict between the 2 scripts?
Food for thought.

ExtremeFiretop commented:
> > @Martinski4GitHub
> > After some deeper investigation with [Stephen](https://www.snbforums.com/members/stephen-harrington.53476/) from the forums. It seems his reported issue related to Dual-WAN failover actually has nothing to do with our code (directly).
> > I tested by trying to setup one of my second routers in Dual-WAN with fail over and was unable to replicate his reports. Everything worked perfectly for me.
> > So I reached out to him via PM and made him provide me some nvram values, some system logs and run some functions, and it's clear to me that the issue is he is using another script for the Dual-WAN fail-over compared to the stock built-in fail-over function I was using.
> > In my tests everything was rock solid but I was using the built-in fail over function with the firmware, he had that disabled and was using this script: dual-wan-failover found here: https://www.snbforums.com/threads/dual-wan-failover-v2-release.83674/
> > In my tests I get a clear "connected" status in the system logs that triggers for both WAN interfaces as found below:
> > FAIL-OVER to WAN1
> > ```
> > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 init)
> > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 connecting)
> > Sep  8 16:40:37 custom_script: Running /jffs/scripts/service-event-end (args: restart wan_if)
> > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 disconnected)
> > Sep  8 16:40:38 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > Sep  8 16:40:39 custom_script: Running /jffs/scripts/service-event (args: restart wan_line)
> > Sep  8 16:40:39 custom_script: Running /jffs/scripts/wan-event (args: 1 connected)
> > Sep  8 16:40:39 custom_script: Running /jffs/scripts/firewall-start (args: eth4)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > FALL-BACK to WAN0
> > ```
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 init)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 connecting)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/service-event-end (args: restart wan_if)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 disconnected)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 connected)
> > Sep  8 18:49:29 custom_script: Running /jffs/scripts/firewall-start (args: eth0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However Stephen with the other script is only getting one "connected" state for WAN0 and nothing for WAN1.
> 
> If he's not getting the built-in, F/W-provided WAN events for the 2nd WAN interface (`wan1`), then all bets are off because the current implementation relies on the built-in WAN "**connected**" events to trigger the WAN status changes which, in turn, trigger the update for the WAN uptime (for those routers that don't have the NVRAM value keeping the uptime by itself).
> 
> The fact that he has the native F/W DUAL-WAN functionality **disabled** it's another red flag indicating that we're not dealing with a standard WAN interface and with the expected F/W notifications for WAN events.
> 
> > He also re-confirmed the issue exists with the current production version now.. This goes against his original report found in the forums here to you confirming it worked correctly: https://www.snbforums.com/threads/scmerlin-2-5-41-service-and-script-control-menu-for-asuswrt-merlin.89224/post-963668
> > So in summery it seems like this issue exists in production and was not introduced in the dev branch, I can confirm that all works correctly with the stock firmware fail-over and so can Stephen.
> > So the question now is, is it worth troubleshooting further or call it a known issue/compatible conflict between the 2 scripts? Food for thought.
> 
> IMO, there's really nothing else for us to do to fix the behavior. Once another 3rd-party script is introduced in the mix, which changes the standard F/W functionality, and if the expected WAN events are suppressed or disabled, we're no longer in known territory with known rules. If we try to do anything to "fix" it, it could break at any time later on if/when the other script decides to change the "rules" again.

The remaining question would be: Is it the script itself that is creating a "weird" fail-over all under the "wan0" interface/event, or is this normal behavior for there to not be a "wan1" even in stock firmware if you have "Allow failback" disabled.

<img width="1055" height="851" alt="image" src="https://github.com/user-attachments/assets/2f6407c8-016c-4ddb-bc6f-49a98bb45eb1" />

ExtremeFiretop commented:
This reason being his script makes having that disabled a prerequisite.
While in all my testing I have it enabled.

Martinski4GitHub commented:
> > > @Martinski4GitHub
> > > After some deeper investigation with [Stephen](https://www.snbforums.com/members/stephen-harrington.53476/) from the forums. It seems his reported issue related to Dual-WAN failover actually has nothing to do with our code (directly).
> > > I tested by trying to setup one of my second routers in Dual-WAN with fail over and was unable to replicate his reports. Everything worked perfectly for me.
> > > So I reached out to him via PM and made him provide me some nvram values, some system logs and run some functions, and it's clear to me that the issue is he is using another script for the Dual-WAN fail-over compared to the stock built-in fail-over function I was using.
> > > In my tests everything was rock solid but I was using the built-in fail over function with the firmware, he had that disabled and was using this script: dual-wan-failover found here: https://www.snbforums.com/threads/dual-wan-failover-v2-release.83674/
> > > In my tests I get a clear "connected" status in the system logs that triggers for both WAN interfaces as found below:
> > > FAIL-OVER to WAN1
> > > ```
> > > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 init)
> > > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 connecting)
> > > Sep  8 16:40:37 custom_script: Running /jffs/scripts/service-event-end (args: restart wan_if)
> > > Sep  8 16:40:37 custom_script: Running /jffs/scripts/wan-event (args: 0 disconnected)
> > > Sep  8 16:40:38 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > > Sep  8 16:40:39 custom_script: Running /jffs/scripts/service-event (args: restart wan_line)
> > > Sep  8 16:40:39 custom_script: Running /jffs/scripts/wan-event (args: 1 connected)
> > > Sep  8 16:40:39 custom_script: Running /jffs/scripts/firewall-start (args: eth4)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > FALL-BACK to WAN0
> > > ```
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 init)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 connecting)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/service-event-end (args: restart wan_if)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 disconnected)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 stopped)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/wan-event (args: 0 connected)
> > > Sep  8 18:49:29 custom_script: Running /jffs/scripts/firewall-start (args: eth0)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > However Stephen with the other script is only getting one "connected" state for WAN0 and nothing for WAN1.
> > 
> > 
> > If he's not getting the built-in, F/W-provided WAN events for the 2nd WAN interface (`wan1`), then all bets are off because the current implementation relies on the built-in WAN "**connected**" events to trigger the WAN status changes which, in turn, trigger the update for the WAN uptime (for those routers that don't have the NVRAM value keeping the uptime by itself).
> > The fact that he has the native F/W DUAL-WAN functionality **disabled** it's another red flag indicating that we're not dealing with a standard WAN interface and with the expected F/W notifications for WAN events.
> > > He also re-confirmed the issue exists with the current production version now.. This goes against his original report found in the forums here to you confirming it worked correctly: https://www.snbforums.com/threads/scmerlin-2-5-41-service-and-script-control-menu-for-asuswrt-merlin.89224/post-963668
> > > So in summery it seems like this issue exists in production and was not introduced in the dev branch, I can confirm that all works correctly with the stock firmware fail-over and so can Stephen.
> > > So the question now is, is it worth troubleshooting further or call it a known issue/compatible conflict between the 2 scripts? Food for thought.
> > 
> > 
> > IMO, there's really nothing else for us to do to fix the behavior. Once another 3rd-party script is introduced in the mix, which changes the standard F/W functionality, and if the expected WAN events are suppressed or disabled, we're no longer in known territory with known rules. If we try to do anything to "fix" it, it could break at any time later on if/when the other script decides to change the "rules" again.
> 
> The remaining question would be: Is it the script itself that is creating a "weird" fail-over all under the "wan0" interface/event, or is this normal behavior for there to not be a "wan1" event in stock firmware if you have "Allow failback" disabled.
> <img alt="image" width="1055" height="851" src="https://private-user-images.githubusercontent.com/1971404/487252498-2f6407c8-016c-4ddb-bc6f-49a98bb45eb1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc0MTY4MzMsIm5iZiI6MTc1NzQxNjUzMywicGF0aCI6Ii8xOTcxNDA0LzQ4NzI1MjQ5OC0yZjY0MDdjOC0wMTZjLTRkZGItYmM2Zi00OWE5OGJiNDVlYjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDlUMTExNTMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzNiNTkzOTQxMTE3ZjdlYjlmMDM5ZmNkMWI3YWI3MTA3ZjE1NzQ0NzVjYmRhODQ4ZjRmNWE3ZDFlMzI5OTU4OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.FvS1Rvro_ZAFnWM81NlCT0Fkh3EmU1JEeL3WqKfKFsw">
>
>
>
Good question! I don't have any way to test this scenario myself. I suppose it's possible that it could be a bug in the F/W. Based on the SNB Forums posts that I've read over the years, the built-in DUAL-WAN functionality is known to have several issues, so I would not be surprised if that's another bug.

Regardless of where the problem is WRT not getting WAN events for the `wan1` interface, that is a show stopper for our current implementation. You could investigate the problem further if you have the time and the desire to pursue it, of course.

P.S.
Going offline now. I have a meeting at 10AM, so I'd better get some sleep to be clear-minded and refreshed.
Talk to you later.

Martinski4GitHub commented on scmerlin.sh:
```diff
@@ -1843,32 +1843,19 @@ Get_WAN_Uptime_JS()
 
-    if [ -n "$rawxStr" ] && echo "$rawxStr" | grep -qiE '^WAN[0-1]:.*'
+    # Normalize whitespace; parse only exact labels #
+    if [ -n "$rawxStr" ]
```
> I am testing a bug causing a "bug" to return "WAN is down" for only a moment before incrementing the counter even though the cable is unplugged. It can eventually be put back but I am not finished yet.
> 
> The goal was to remove this stricter validation to see if it was the one saying the "WAN is down" or if it was from within the Get_WAN_Uptime function since they both output that messaging depending on the interface online, and I needed to narrow down which function was causing it.
>
>
OK, got it. It was not clear at all whether the change was temporary for testing/debugging purposes only, or if it was meant to be permanent. Usually, one would indicate the intention with a comment line inserted before the modified code block or line. For example:
```
## MODIFIED FOR TESTING/DEBUGGING PURPOSES ONLY ##
```
In any case, it looks like you have been busy working behind the scenes with Stephen to get things working with his setup. Hopefully, you have reached a good solution at this point.

P.S. 
This past week, I was very busy with work, and I'll be starting a new project with a new team starting the 1st week of October, so I may be offline more often and may not reply within 24 to 48 hours, as I usually try to do.

We just came back home visiting my in-laws. Over the weekend, I finally convinced my own folks and my in-laws to replace all their Windows 10 laptops with Windows 11 due to the approaching October 15th deadline. So we got new laptops and I set them all up for them, so it was pretty much painless and transparent for their usage.

I'm not sure what to do with all the old Windows 10 gear (4 laptops). So much electronic waste, but they're old (7 to 9-year-old equipment), and I don't want to spend money and time putting Linux OS or upgrading components.

Anyway, I hope you had a good weekend. Talk to you whenever I get some free time.
I'll work on YazDHCP for a couple of hours before going to bed. I'd like to finish the new feature before the end of this month.

ExtremeFiretop replied:
@Martinski4GitHub 

Very nice of you to go through the trouble of upgrading all their devices.
I know the problem of electronic waste is a real shame when a lot of these devices still work without issue with the older OS.

Like you said you can usually put Linux on them, but that is time and maintenance. Although Linux distro like Ubuntu these days are extremely user friendly compared to the early days.

I've been busy myself with personal life things. But I think I've nailed down this Dual-WAN issue in Load Balancing. The issue is the nvram values we were using didn't seem to be consistent or reliable. Instead in my latest commits I've swapped to some new values used by the other dual -wan script with much more solid success so far.

Will do a bit more testing to make sure I don't cause any regression issues but so far so good!

Martinski4GitHub replied:
> @Martinski4GitHub
> 
> Very nice of you to go through the trouble of upgrading all their devices. I know the problem of electronic waste is a real shame when a lot of these devices still work without issue with the older OS.
> 
> Like you said you can usually put Linux on them, but that is time and maintenance. Although Linux distro like Ubuntu these days are extremely user friendly compared to the early days.
> 
> I've been busy myself with personal life things. But I think I've nailed down this Dual-WAN issue in Load Balancing. The issue is the nvram values we were using didn't seem to be consistent or reliable. Instead in my latest commits I've swapped to some new values used by the other dual -wan script with much more solid success so far.
> 
>
>
OK, sounds like you have a good handle on the Dual-WAN with Load Balancing setup. I remember sometime last year, I took a look at the Dual-WAN script, and it was really awful to read. I'm guessing the author is not a professional s/w developer, and he's doing the best he can, so it's not really fair to judge; and to his credit, the script seems to be doing its job well enough.

> 
> Will do a bit more testing to make sure I don't cause any regression issues but so far so good!
>
Yeah, that sounds like a good plan. We want to avoid regression bugs, especially with other users who have the built-in, native Dual-WAN setups.

Have a good night, bud. Going to bed myself in a couple of minutes.

Martinski4GitHub commented:
> @Martinski4GitHub
> 
> I think I've solved both reported issues with Dual-WAN. (With Fail-over mode not triggering a wan-event and with Load-Balancing mode not flipping the expected nvram _state_t values)
>
>
OK, very good. It took much longer than anticipated, I'm sure, but it looks like you got it in the bag now!! LOL!!

> 
> I'm having both users re-test and just sent a message to them tonight, but my results were perfect. Can I ask you to do a review of the final changes and quiz me on anything in particular? I'm ready for your review now :)
>
>
I'll take a look at it sometime tomorrow evening or over the weekend. Right now, I'm trying to finish the new YazDHCP feature and make it ready for a Beta release in the next few days. I'll let you know when I'm done with the review of this PR.

Martinski4GitHub commented:
@ExtremeFiretop,

I see that you've made several more commits since last time we talked, so testing appears to be still in progress. I'll wait until you get final feedback with all your latest changes and you feel confident that you got the solution nailed down. I'm not in a hurry at all, so take your time.

Have a good night, bud!!!

ExtremeFiretop commented:
> @ExtremeFiretop,
> 
> I see that you've made several more commits since last time we talked, so testing appears to be still in progress. I'll wait until you get final feedback with all your latest changes and you feel confident that you got the solution nailed down. I'm not in a hurry at all, so take your time.
> 
> Have a good night, bud!!!

Thank you my friend, you are correct as always.

Essentially, I waited a week for [Tarek Yag](https://www.snbforums.com/members/tarek-yag.99657/) to come back on the forums and give me some feedback, I don't mind the long delays because as we creep up to the 1st I have less and less time, I may want to put an out of office on or something for the forums. I am not 4 days away from getting keys to the new place!

His latest feedback was that it worked in all instances except one, let me condense down all the feedback he gave to the important points:


```
Yep, they all events fired up accordingly and as expected, except for the last test when I plugged the WAN0 cable again, no proper WAN events!

See the WAN log at:
Sep 22 04:40:39 ATR wan: finish adding multi routes
Sep 22 04:40:39 ATR WAN(0)_Connection: WAN was restored.

Yes, again, everything worked exactly as expected, except for the last test when I plugged WAN0 cable again, I got no WAN events at all, and the uptime counter continued counting right from where it stopped when the cable was plugged out.
```

And:

```
Meanwhile, don't forget that I lose connectivity with WAN0 router even after plugging in the cable again. 
So, I believe it's worthless working around this issue, since the firmware loses WAN0 connectivity altogether!!
```

So the latest changes you saw me commit "Add additional cleanups" which I then reverted was to attempt to "zero" his counters so it would say "WAN is down" instead of "continuing where it left off" even though he had no wan-events, and said the following:

```
ExtremeFiretop says:
And to be clear, the expectation is that is says WAN is down due to this? correct? 
Currently it continues from where it left off? But you would expect it to say it's down since WAN0 loses connectivity, yes?
```

He tested the changes and said:

```
scMerlin now keeps showing WAN0 as "down" with the latest change.

I'd prefer that the counter continues from where it left off.
```

He clarified that infact even though the WebUI and some elements are not accessible over WAN0, that the router is able to ping out WAN0 so therefore it should be considered up, so i reverted the change. AKA last change was no good.

I then clarified that we cannot do much of the router is completely failing to trigger a wan-event:

```
ExtremeFiretop says:
The evidence you provided in the system logs was of no wan-events at all, not that the events were passing the wrong argument or that it's not landing on a "connected" state.
Which means there is little to nothing that can be done, all bets are off if no wan-events are detected at all.
```

And reverted the change.
Here is the last message he sent to me:

```
The load-balance feature is so MESSED UP in the stock firmware!!
Just cancel the latest change you just did, keeping what you've accomplished so far, as I believe it's the best that can be done.

Yep, I guess you can issue a release at your convenience.
I guess this is a dead-end by the stupid firmware I have on my very old device!
```

Martinski4GitHub reviewed (approved):
It's good to go live!!
